source/extensions/filters/network/thrift_proxy/decoder_events.h

```diff
@@ -35,6 +35,20 @@ class DecoderEventHandler {
    */
   virtual FilterStatus transportEnd() PURE;
 
+  /**
+   * @return True if payload passthrough is enabled.
+   * Once any filter returns false, the payload passthrough mode is disabled.
+   */
+  virtual bool passthroughEnabled() const PURE;
```

Comment:
This function is a bit confusing to me. I think the trouble is that in ConnectionManager::ActiveRpc it indicates whether passthrough was configured (and is allowed based on the filter chain), but for filters themselves it indicates whether or not the filter can support passthrough.

I think a way to resolve this is to move `bool passthroughEnabled()` to the DecoderCallbacks interface, and to add a reference to the DecoderCallbacks to Decoder::ActiveRequest alongside the DecoderEventHandler. That allows access to the callbacks from the DecoderStateMachine. Second, add `bool passthroughSupported()` (feel free to pick a different name, just not passthroughEnabled) to ThriftFilters::DecoderFilter to indicate whether a filter itself supports passthrough. The contrat of that method is that it's called once, right `messageBegin`. That allows allows moving the enabled bool out of ProtocolConverter and keeping it in the Router, or perhaps just computing the result at the time the method is called.

Provided it all the visibility works out, I think that makes clearer which method does what. What do you think?

Reply:
Sounds very reasonable, just to make sure i understand correctly, does this mean `bool passthroughSupported()` should also be added to `DecoderEventHandler`, and DecoderStateMachine checks `decoder_callbacks_.passthroughEnabled ()` and `handler_.passthroughSupported()` in `messageBegin` ?

Reply:
I think the `passThroughSupported` goes in the DecoderFilter interface (in `source/extensions/filters/network/thrift_proxy/filters/filter.h`). So filters indicate whether they support pass through and the DecoderCallbacks can tell the state machine whether or not passthrough is in use (enabled).

DecoderEventHandler just gets the `passthroughData` method.

Reply:
I understand the changes in the interface. I'm trying to figure out how `DecoderEventHandler` knows filter chain support passthrough.  `DecoderEventHandler` calls `decoder_callbacks_.passthroughEnabled()`  In `messageBegin`,  the result indicates whether passthrough was configured (and is allowed based on the filter chain). `ConnectionManager` implements `passthroughEnabled` by checking config option and `rpcs_.begin()`, then `ActiveRpc` calls `passThroughSupported` on every filter. Is that right ?

Reply:
updated